### ISSUE TITLE
Object.BlockingHeight: Feature design is broken

### ISSUE THREAD

**Issue by fernewelten** [adventuregamestudio/ags#1413]
Have a look at the following diagram:

![BugReport1](https://user-images.githubusercontent.com/41213694/133279882-e57e21cb-be6b-467e-b95f-3f39a63281d0.png)

The walking area is in light blue.
`Object.BlockingHeight` in combination with `Object.Baseline` is a feature that is supposed to stop characters from walking through the object. This is accomplished by cutting out a rectangular area out of the walking area

So let's put a small cabinet into the room and see, from a purely geometric vantage point, what area would need to be cut out. It would be an area that is delimited to the left and right by the outer edges of the cabinet (violet lines in the pic), to the bottom by the baseline (blue) and to the top by whatever line the back edge of the cabinet works out to (green line).

Verify, using the persons in the picture, that they won't walk through the cabinet as long as they stay below the blue line or above the green line.

From a purely geometric standpoint, the blocking area lies completely above the baseline. From there, the blocking height should extend upwards until the back of the cabinet is reached. That is the area between the blue and green and violet lines. Of course, from a practical standpoint, a small slice of the walking area __below__ the baseline should be cut out, too, and also a small slice of the walking area __above__ where the back of the cabinet works out to.  

But that's not at all what we are doing with our `Object.Baseline` concept. We're cutting out an area that extends __symmetrically__ around the baseline. 

This isn't going to work, even from a practical vantage point: If we extend the blocked area high enough so that the back of the cabinet is reached, then the area below the cabinet will become far too large. If we make the blocked area 5 pixels below and above the baseline (which is the default, probably for the 320x200 screens that AGS was originally designed for) then the blocked area doesn't extend high enough by far to prevent characters from walking through the object.

Basically, it's __wrong__ to define the blocking area to be symmetric around the baseline. The __design__ of the feature is broken.

I think that this is the reason that there are complaints on the website that "blocking doesn't work" and that coders should prefer  keeping out Ego by defining suitable walking areas.

How can we improve that?

**Comment by ivan-mogilko** [comment]
Personally, I believe that the only reliable way to cover all the use cases is to have a "collider" component defined by:
* a mask (default to same rectangle)
* collider's offset counted from the object's "origin point" (this is where it's coordinates are at).

**Comment by fernewelten** [comment]
If we let the coders define a "stencil" shape for objects, then we'd basically have two cases: 
- A game that has violet as the 'invisible' colour (in this case the "stencil" cuts out all pixels that are anything but the 'invisible' colour)
- A game with an 8-bit alpha channel (we could define a fixed constant for that case that has to be reached so that the stencil will cut off, e.g., all the pixels where the alpha channel of the stencil is 128 or higher). 
Call this `BlockingStencil` or some such, which would be a sprite number?

If we keep to a rectangle then IMO the main need is to be able to specify an unsymmetric area around the **baseline**. It's okay to have a symmetric margin around the object's/character's outer sides..

So perhaps keep `.BlockingWidth` but split up `.BlockingHeight` into `.BlockingHeightAboveBaseline` and `.BlockingHeightBelowBaseline`? But those names are quite a mouthful. Or keep `BlockingHeight` symmetric, but not along the baseline but along some other line that the coder may specify? `BlockingLine` or some such?

I think that solidness would be a very useful feature when it works. I'd probably use it all the time. For instance: So there's a door in the room. When you click on it, it "opens", i.e. an "open door blade" object that is lying on top of the door is made visible. So as of now, of course, nobody may walk through the door blade. 

Or a small cabinet is in the room that can be pushed around. Characters shouldn't walk through the cabinet, no matter where it has been pushed.

**Comment by fernewelten** [comment]
Another idea would be to have separate "cutout rectangles" given by `Left`, `Width`, `Top`, and `Height`, independent from objects and modifiable at runtime. Wherever a "cutout rectangle" comes to lie, it cuts out the walkable area below it. 

If the program can access the graphic that contains the walkable area at runtime as a `DrawingSurface´, it could cut out sections with the usual `DrawingSurface` commands. But this would perhaps have ramifications for `Object.Move()` and `Character.Walk()`.

**Comment by ivan-mogilko** [comment]
> If we let the coders define a "stencil" shape for objects, then we'd basically have two cases:

Guess I was not clear with a "mask", I did not mean something that cuts the object's sprite, I meant something that defines the blocking area, and used for hit tests. In such case object's colour depth is irrelevant, and the mask could be even 1-bit map.

> So perhaps keep `.BlockingWidth` but split up `.BlockingHeight` into `.BlockingHeightAboveBaseline` and `.BlockingHeightBelowBaseline`? But those names are quite a mouthful. Or keep `BlockingHeight` symmetric, but not along the baseline but along some other line that the coder may specify? `BlockingLine` or some such?

Should not having blocking area "offset" cover this? as you may offset it anywhere, in any direction both vertically and horizontally.

**Comment by ivan-mogilko** [comment]
@fernewelten , do you think that recently added BlockingRectX and BlockingRectY solve this issue?
https://adventuregamestudio.github.io/ags-manual/v3/Character.html#characterblockingrectx

**Comment by ivan-mogilko** [comment]
I think I'll close this as fixed by BlockingRectX, BlockingRectY properties, since they help solving the mentioned problem:
> Basically, it's wrong to define the blocking area to be symmetric around the baseline. The design of the feature is broken.

However, there's still a question about redesigning the whole collision test mechanic, e.g. with the optional use of collision masks (mentioned in one of the comments above). If anyone's interested, that could be opened as a separate feature suggestion.
